### PR TITLE
Test to show collections and lists of references work (& fix for lists)

### PR DESCRIPTION
**Pull request by shans**
As described in b/160556762, it looked like collections and lists of References would not correctly be injected with a Dereferencer.

This patch confirms that collections *are* handled correctly, demonstrates that lists *weren't*, and fixes the EntityDereferencerFactory logic for the list case.

The overloaded (private) injectDereferencers has also been renamed for clarity, and the inner function injectField broken out as a separate method. This may constitute sufficient clean-up to render the code in EntityDereferencerFactory clear - I know what it does now so I can't tell any more..

### CODE REVIEW

File: java/arcs/core/entity/EntityDereferencerFactory.kt [hunk 1]
```diff
@@ -32,16 +33,21 @@ class EntityDereferencerFactory(
...
 
-    private fun injectDereferencers(schema: Schema, rawEntity: RawEntity) {
+    private fun injectDereferencersIntoRawEntity(schema: Schema, rawEntity: RawEntity) {
         fun injectField(fieldType: FieldType?, fieldValue: Any?) {
             val schemaHash = when (fieldType) {
                 is FieldType.EntityRef -> fieldType.schemaHash
                 is FieldType.InlineEntity -> fieldType.schemaHash
+                is FieldType.ListOf -> {
+                    injectField(fieldType.primitiveType, fieldValue)
+                    null
```
Comment (galganif) [inline java/arcs/core/entity/EntityDereferencerFactory.kt:49]: injectField returns null, can you omit this line?

File: java/arcs/core/entity/EntityDereferencerFactory.kt [hunk 2]
```diff
@@ -32,16 +33,21 @@ class EntityDereferencerFactory(
...
-            is RawEntity -> injectDereferencers(schema, value)
+            is RawEntity -> injectDereferencersIntoRawEntity(schema, value)
             is Set<*> -> value.forEach { injectDereferencers(schema, it) }
+            is ReferencableList<*> -> value.value.forEach { injectDereferencers(schema, it) }
         }
     }
 
-    private fun injectDereferencers(schema: Schema, rawEntity: RawEntity) {
+    private fun injectDereferencersIntoRawEntity(schema: Schema, rawEntity: RawEntity) {
         fun injectField(fieldType: FieldType?, fieldValue: Any?) {
```
Comment (galganif) [inline java/arcs/core/entity/EntityDereferencerFactory.kt:43]: as another readability improvement, what do you think about moving injectField to be a a separate top level private function?

File: javatests/arcs/core/entity/HandleManagerTestBase.kt [hunk 3]
```diff
@@ -734,10 +734,59 @@ open class HandleManagerTestBase {
...
+            entitySpec = TestReferencesParticle_Entities
+        )
+        val readHandle = readHandleManager.createCollectionHandle(
+            entitySpec = TestReferencesParticle_Entities
+        )
+
+        val updateDeferred = readHandle.onUpdateDeferred { it.size == 3 }
+        entities.forEach { writeHandle.dispatchStore(it) }
+        val entitiesOut = updateDeferred.await()
+        assertThat(entitiesOut).containsExactly(*entities.toTypedArray())
```
Comment (galganif) [inline javatests/arcs/core/entity/HandleManagerTestBase.kt:783]: does `assertThat(entitiesOut).containsExactlyElementsIn​(entities)` work?

File: java/arcs/core/entity/EntityDereferencerFactory.kt (hunk 1 above)
Comment (shans) [reply java/arcs/core/entity/EntityDereferencerFactory.kt:49]: I worry that it makes the code flow significantly less clear if I do. This is kind of documentation more than anything. WDYT?

Comment (galganif) [reply java/arcs/core/entity/EntityDereferencerFactory.kt:49]: sounds good

File: java/arcs/core/entity/EntityDereferencerFactory.kt (hunk 2 above)
Comment (shans) [reply java/arcs/core/entity/EntityDereferencerFactory.kt:43]: Yup, good idea!

File: javatests/arcs/core/entity/HandleManagerTestBase.kt (hunk 3 above)
Comment (shans) [reply javatests/arcs/core/entity/HandleManagerTestBase.kt:783]: it does, thanks.

**Pull request merged**